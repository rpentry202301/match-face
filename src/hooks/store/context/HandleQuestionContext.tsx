"use client";

import React, { useContext } from "react";

type Refine = {
  word: string[];
  department: number[];
};

const SelectContext = React.createContext<[Refine, React.Dispatch<React.SetStateAction<Refine>>] | undefined>(undefined);

export function SelectProvider({ children }: { children: React.ReactNode }) {
  const [select, setSelect] = React.useState<Refine>({
    word: [],
    department: [],
  });

  return (
    <SelectContext.Provider value={[select, setSelect]}>
      {children}
    </SelectContext.Provider>
  );
}

export function useRefine() {
  const context = useContext(SelectContext);
  if (context === undefined) {
    throw new Error("useRefine must be used within a SelectProvider");
  }
  return context;
}
