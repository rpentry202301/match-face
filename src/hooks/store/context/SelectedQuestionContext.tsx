"use client";
import {
  Dispatch,
  ReactNode,
  Reducer,
  createContext,
  useContext,
  useReducer,
} from "react";
import { QusetionsReducer } from "../reducer/selecterReducer";
import { QuestionsReducerAction } from "@/types/admin/tasks/register/types";

export type Questions = {
  projectId: number;
  list: {
    id: number;
    name: string;
  }[];
};

const SelectedQuestionContext = createContext<
  [Questions, Dispatch<QuestionsReducerAction>] | undefined
>(undefined);

export const SelectedQuestionProvider = ({
  children,
}: {
  children: ReactNode;
}) => {
  const [state, dispatch] = useReducer<
    Reducer<Questions, QuestionsReducerAction>
  >(QusetionsReducer, { projectId: 0, list: [] });
  return (
    <SelectedQuestionContext.Provider value={[state, dispatch]}>
      {children}
    </SelectedQuestionContext.Provider>
  );
};

export const useSelectedQuestion = () => {
  const context = useContext(SelectedQuestionContext);
  if (context === undefined)
    throw new Error("Function 'useSelectedQuestion' must be used in Provider.");
  return context;
};
