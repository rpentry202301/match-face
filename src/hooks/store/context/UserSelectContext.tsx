"use client";
import {
  Dispatch,
  ReactNode,
  Reducer,
  createContext,
  useContext,
  useReducer,
} from "react";
import { UsersReducer } from "../reducer/selecterReducer";
import { UsersReducerAction } from "@/types/admin/tasks/register/types";

export type Users = {
  id: string;
  name: string;
};

const UserSelectContext = createContext<
  [Users[], Dispatch<UsersReducerAction>] | undefined
>(undefined);

export const UserSelectProvider = ({ children }: { children: ReactNode }) => {
  const [state, dispatch] = useReducer<Reducer<Users[], UsersReducerAction>>(
    // selecterReducer,
    UsersReducer,
    []
  );
  return (
    <UserSelectContext.Provider value={[state, dispatch]}>
      {children}
    </UserSelectContext.Provider>
  );
};

export const useUserSelect = () => {
  const context = useContext(UserSelectContext);
  if (context === undefined)
    throw new Error("Function 'useUserSelect' must be used in Provider.");
  return context;
};
