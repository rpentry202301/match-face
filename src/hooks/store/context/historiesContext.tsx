'use client'
import { createContext, Dispatch, useContext, useReducer } from "react";
import { SelectHistoryAction } from "@/types/admin/histories/admin_histories";
import { SelectHistoryReducer } from "../reducer/historiesReducer";

const SelectHistoryContext = createContext<{month:string,department:number[],skills:number[]}| undefined>(undefined);
const SelectHistoryDispatchContext = createContext<Dispatch<SelectHistoryAction>|undefined>(undefined)
const initialSelect = {month:"",department:[],skills:[]}

export function SelectHistoryProvider ({children}:{children:React.ReactNode}){
    const [state,dispatch] = useReducer(SelectHistoryReducer,initialSelect)
    return (
        <SelectHistoryContext.Provider value={state}>
            <SelectHistoryDispatchContext.Provider value={dispatch} >
            {children}
            </SelectHistoryDispatchContext.Provider>
        </SelectHistoryContext.Provider>
    )
}

export const useSelectHistory = () => {
    const context = useContext(SelectHistoryContext)
    if(context === undefined) {
        throw new Error("Function 'useSelectHistory' must be used in Provider.")
    }
    return context
}

export const useSelectHistoryDispatch = () => {
    const dispatch = useContext(SelectHistoryDispatchContext)
    if(dispatch === undefined) {
        throw new Error("Function 'useSelectHistoryDispatch' must be used in Provider.")
    }
    return dispatch
}
