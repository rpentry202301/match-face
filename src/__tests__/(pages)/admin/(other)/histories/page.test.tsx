import { act, render,screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import HistoriesPage from '@/app/(pages)/admin/(other)/histories/page'
import HistoriesSelect from '@/components/pages/admin/histories/select'
import React from 'react'
import "@testing-library/jest-dom"
import { AppRouterContextProviderMock } from '@/__tests__/test_utils/app-router-context-provider-mock'
import HistoriesList from '@/components/pages/admin/histories/list'
import { SelectHistoryProvider } from '@/hooks/store/context/historiesContext'
import { skillsMock, departmentsMock,answer_request_groupsMock } from './histories_test_mock'

describe('管理者/履歴一覧画面のテスト',() => {
    const user = userEvent.setup()
    const push = jest.fn()
    global.fetch = jest.fn().mockImplementation((url,config)=>{
        if(url === 'http://localhost:3000/api/admin/histories/departments'){
            return {
                ok:true,
                json: async () => (departmentsMock),
            }
        }else if(url === 'http://localhost:3000/api/admin/histories/skills'){
            return {
                ok:true,
                json: async () => (skillsMock),
            }
        }else if(url === 'http://localhost:3000/api/admin/histories'){
            return {
                ok:true,
                json: async () => (answer_request_groupsMock),
            }
        }
    })
    describe('スナップショットテスト',() => {
        it('レンダリング時',async() => {
            const view = render(
                <SelectHistoryProvider>
                    <HistoriesPage/>
                </SelectHistoryProvider>
            )
            await waitFor(()=>{
                expect(view.container).toMatchSnapshot()
            })
        })
    })
    describe('リストテスト',()=>{
        beforeEach(async()=>{
            await act(()=>render(
                <SelectHistoryProvider>
                    <AppRouterContextProviderMock router={{ push }}>
                        <HistoriesSelect className={''}/>
                        <HistoriesList />
                    </AppRouterContextProviderMock>
                </SelectHistoryProvider>
            ));
        })
        it('オープンボタンを押すとクローズボタンとユーザーリストが表示される',async()=>{
            const openButton = screen.getByTestId('open_1');
            await user.click(openButton);
            const closeButton = screen.getByTestId('close_1');
            const userList = screen.getByText('氏名')
            expect(closeButton).toBeInTheDocument();
            expect(userList).toBeInTheDocument();
            expect(openButton).not.toBeInTheDocument();
        })
        it('詳細ボタンを押すとページ遷移する',async()=>{
            const openButton = screen.getByTestId('open_1')
            await (user.click(openButton))
            const detailButton = screen.getByTestId('detail_1')
            expect(detailButton).toBeTruthy()
            await user.click(detailButton)
            expect(push).toBeCalled()
            expect(push).toBeCalledWith('/admin/review/1')
        })
    })
    describe('HistoriesListコンポーネントテスト',()=>{
        beforeEach(async()=>{
            await act(()=>render(
                <SelectHistoryProvider>
                <AppRouterContextProviderMock router={{ push }}>
                    <HistoriesList />
                </AppRouterContextProviderMock>
                </SelectHistoryProvider>
            ))
        })
        it('説明文の長さに合わせて表示が変わる',async()=>{
            await waitFor(async()=>{
                expect(screen.getByTestId('project_detail_2').textContent).toBe('テスト2の説明(文字数34)テスト2の説明テスト2の説明テスト2の説明');
                expect(screen.getByTestId('project_detail_3').textContent).toBe('テスト3の説明(文字数35)テスト3の説明テスト3の説明テスト3の説明...');
            })
            })
        it('回答の状態に合わせて表示が変わる',async()=>{
            await waitFor(async()=>{
                const openButton = screen.getByTestId('open_1')
                await user.click(openButton)
                expect(screen.getByTestId('status_1').textContent).toBe('回答済み')
                expect(screen.getByTestId('status_2').textContent).toBe('未回答')
            })
        })
    })
})
   