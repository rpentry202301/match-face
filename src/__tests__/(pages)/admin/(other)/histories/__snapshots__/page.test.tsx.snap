// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`管理者/履歴一覧画面のテスト スナップショットテスト レンダリング時 1`] = `
<div>
  <section
    class="w-[75vw] ml-[12.5vw] border-2 text-center my-[5vh] py-[5vh] max-md:w-[90vw] max-md:ml-[5vw]"
  >
    <form
      action="submit"
      data-testid="form"
    >
      <div
        class="mb-[2vh]"
      >
        <label
          for="month"
        >
          更新月：
        </label>
        <select
          class="border-2"
          data-testid="month"
          id="month"
          name="month"
        >
          <option
            value=""
          >
            --
          </option>
        </select>
      </div>
      <div
        class="mb-[1.5vh] flex mx-auto flex-wrap w-[65vw] justify-between max-lg:after:contents:'' max-lg:after:w-[100px] max-lg:before:contents:'' max-lg:before:w-[100px] max-lg:before:order-1"
      />
      <div
        class="lg:flex justify-center mb-[2vh] w-[55vw] ml-[10vw] max-md:w-[70vw]"
      >
        <legend
          class="lg:w-[215px] lg:h-10 lg:leading-10"
        >
          使用技術：
        </legend>
        <fieldset
          class="text-justify"
          id="skill"
          name="skill"
        />
      </div>
      <div>
        <button
          class="bg-orange drop-shadow-md hover:saturate-150 active:drop-shadow-none active:shadow-inner hover:bg-depp-orange active:bg-depp-orange active:translate-y-0.5 text-white rounded-xl p-2 w-40 h-auto text-lg"
          type="submit"
        >
          絞り込み
        </button>
      </div>
    </form>
  </section>
  <section
    class=" max-md:overflow-x-auto max-md:mb-[30px]"
  >
    <table
      class="border-collapse border border-slate-deep-gray w-[75vw] ml-[12.5vw] text-center mb-[10vh] max-md:w-[180vw] max-md:mb-1 max-md:mx-[5vw]"
    >
      <thead>
        <tr>
          <th
            class="border border-slate-deep-gray bg-light-gray w-[5vw] py-2"
            id="open"
          />
          <th
            class="border border-slate-deep-gray bg-light-gray w-[15vw] py-2"
            id="answer_answerDate"
          >
            最終更新日
          </th>
          <th
            class="border border-slate-deep-gray bg-light-gray w-[15vw] py-2"
            id="project_name"
          >
            案件名
          </th>
          <th
            class="border border-slate-deep-gray bg-light-gray w-[40vw] py-2"
            id="project_detail"
          >
            案件概要
          </th>
        </tr>
      </thead>
    </table>
  </section>
</div>
`;
